benchmark/client/relay.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 PairMesh, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/flynn/noise"
+	"github.com/pairmesh/pairmesh/benchmark/config"
+	"github.com/pairmesh/pairmesh/benchmark/results"
+	"github.com/pairmesh/pairmesh/benchmark/utils"
+	"github.com/pairmesh/pairmesh/internal/relay"
+	"github.com/pairmesh/pairmesh/message"
+	"github.com/pairmesh/pairmesh/protocol"
+	"github.com/pairmesh/pairmesh/security"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+type RelayClient struct {
+	cfg *config.ClientConfig
+}
+
+// NewRelayClient returns RelayClient struct with input config
+func NewRelayClient(cfg *config.ClientConfig) *RelayClient {
+	return &RelayClient{
+		cfg: cfg,
+	}
+}
+
+// Start starts relay test cases from client side
+func (c *RelayClient) Start() error {
+	// Since in benchmark test we don't have a portal to transmit credentials,
+	// we will have to make credentials deterministic so that the relay side and
+	// the client side share the same knowledge
+	rng := utils.NewDetermRng()
```

Comment:
I think we can hard code two keys for the benchmark program to make them totally deterministically.

Reply:
I would prefer using this deterministic rng, for several reasons:

1. A deterministic rng looks just more elegant
1. We actually need to hard code 3 things: `priv` key, `serverDHKey` pair and `clientDHKey` pair. All of them are not simple byte arrays, actually they are all structs with internal data structures. Take priv as example, it is a 512 bits thing represented as:
```
// A PrivateKey represents an RSA key
type PrivateKey struct {
	PublicKey            // public part.
	D         *big.Int   // private exponent
	Primes    []*big.Int // prime factors of N, has >= 2 elements.

	// Precomputed contains precomputed values that speed up private
	// operations, if available.
	Precomputed PrecomputedValues
}

```
This makes it really tedious to really hard code everything
3. The generation of RSA itself requires time. Looking at the source code under the hood, it takes actually multiple rounds of rng to find a prime number, etc. In the future if we want to take the RSA generation as a part of benchmark, we also need to mock the actual process of RSA, rather than just a plain hard code.